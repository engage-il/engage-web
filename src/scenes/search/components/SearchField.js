// @flow
import React, { Component } from 'react'
import FontAwesome from 'react-fontawesome'
import { css } from 'glamor'
import { stylesheet, colors, mixins } from 'shared/styles'

export class SearchField extends Component {
  props: {
    value: string,
    styles?: Object,
    onChange: (string) => void
  }

  state = {
    isFocused: false
  }

  // events
  inputDidChange = (event: { target: { value: string } }) => {
    const { onChange } = this.props
    onChange(event.target.value)
  }

  inputDidChangeFocus = (isFocused: boolean) => {
    this.setState({ isFocused })
  }

  // lifecycle
  render () {
    const { value, styles } = this.props
    const { isFocused } = this.state

    return <div {...css(rules.container, styles)}>
      <div {...rules.background} />
      <h1>What's important to you?</h1>
      <div {...css(rules.field, isFocused && rules.focused)}>
        <FontAwesome name='search' />
        <input
          {...rules.input}
          type='text'
          name='search-field'
          value={value}
          placeholder={`health care, HB2364`}
          onChange={this.inputDidChange}
          onFocus={() => this.inputDidChangeFocus(true)}
          onBlur={() => this.inputDidChangeFocus(false)} />
      </div>
    </div>
  }
}

const rules = stylesheet({
  container: {
    display: 'flex',
    flexDirection: 'column',
    '> h1': {
      marginBottom: 15
    }
  },
  field: {
    ...mixins.shadows.low(),
    ...mixins.borders.low(),
    display: 'flex',
    alignItems: 'center',
    maxWidth: 700,
    height: 50,
    padding: '0 15px',
    backgroundColor: colors.neutral,
    transition: 'box-shadow 0.25s, border 0.25s, transform 0.25s',
    transform: 'perspective(200px) translateZ(0px)',
    cursor: 'text',
    ' .fa': {
      width: 15,
      marginRight: 15,
      transition: 'color 0.25'
    },
    ...mixins.mobile({
      width: 'auto'
    })
  },
  focused: {
    ...mixins.shadows.high(),
    ...mixins.borders.high(),
    transform: 'perspective(200px) translateZ(1px)',
    ' .fa': {
      color: colors.primary
    }
  },
  input: {
    ...mixins.fonts.regular,
    flex: 1,
    fontSize: 20,
    border: 'none',
    '::placeholder': {
      color: '#DEDEDE'
    },
    ':focus': {
      outline: 'none'
    }
  }
})
