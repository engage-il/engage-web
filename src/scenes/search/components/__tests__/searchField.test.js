/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { SearchField } from '../SearchField'

// subject
let subject
let onChange

function loadSubject () {
  onChange = jest.fn()
  subject = shallow(<SearchField value={'test value'} onChange={onChange} />)
}

const element = {
  field: () => subject.find('div > div').at(1),
  input: () => subject.find(`input[name='search-field']`),
  icon: () => subject.find('FontAwesome')
}

// specs
afterEach(() => {
  subject = null
})

describe('#state', () => {
  beforeEach(loadSubject)

  it('starts unfocused', () => {
    expect(subject).toHaveState('isFocused', false)
  })
})

describe('#render', () => {
  beforeEach(loadSubject)

  it(`sets in the input's value`, () => {
    expect(element.input()).toHaveValue('test value')
  })

  it('does not have the focused style', () => {
    expect(element.field()).not.toMatchRule(/174vovc/)
  })

  describe(`when it's focused`, () => {
    beforeEach(() => {
      subject.setState({ isFocused: true })
    })

    it('adds the style to the field', () => {
      expect(element.field()).toMatchRule(/i3g58o/)
    })
  })
})

describe('when the value changes', () => {
  beforeEach(() => {
    loadSubject()
    subject.setState({ isFocused: true })
    const event = { target: { value: 'foo' } }
    element.input().prop('onChange')(event)
  })

  it('notifies its parent', () => {
    expect(onChange).toHaveBeenLastCalledWith('foo')
  })
})

describe('when the input is foucsed', () => {
  beforeEach(() => {
    loadSubject()
    element.input().prop('onFocus')()
  })

  it('focuses the component', () => {
    expect(subject).toHaveState('isFocused', true)
  })
})

describe('when the input is blurred', () => {
  beforeEach(() => {
    loadSubject()
    subject.setState({ isFocused: true })
    element.input().prop('onBlur')()
  })

  it('unfocuses the component', () => {
    expect(subject).toHaveState('isFocused', false)
  })
})
