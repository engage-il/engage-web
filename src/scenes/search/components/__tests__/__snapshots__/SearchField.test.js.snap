// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render renders correctly 1`] = `
<div
  data-css-14z4k0p=""
>
  <div />
  <h1>
    What's important to you?
  </h1>
  <div
    data-css-1l4c2h5=""
  >
    <FontAwesome
      name="search"
    />
    <input
      data-css-6sarxw=""
      name="search-field"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="health care, HB2364"
      type="text"
      value="test value"
    />
  </div>
</div>
`;

exports[`#render when it is focused 1`] = `
<div
  data-css-14z4k0p=""
>
  <div />
  <h1>
    What's important to you?
  </h1>
  <div
    data-css-i3g58o=""
  >
    <FontAwesome
      name="search"
    />
    <input
      data-css-6sarxw=""
      name="search-field"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="health care, HB2364"
      type="text"
      value="test value"
    />
  </div>
</div>
`;
